der/src/decoder.rs

```diff
@@ -283,6 +283,29 @@ impl<'a> Decoder<'a> {
         Ok(self.bytes.ok_or(ErrorKind::Failed)?.len())
     }
 
+    /// Obtain a slice of bytes contain a complete TLV production suitable for parsing later.
+    pub fn tlv_slice(&mut self) -> Result<&'a [u8]> {
+        if self.is_failed() {
+            return Err(self.error(ErrorKind::Failed));
+        }
+
+        let pos = self.position();
+        let tag = Tag::decode(self)?;
+
+        let element_length = Length::decode(self).map_err(|e| {
+            if e.kind() == ErrorKind::Overlength {
+                ErrorKind::Length { tag }.into()
+            } else {
+                e
+            }
+        })?;
```

Comment:
I think all of this can be replaced by...

```suggestion
       let header = self.peek_header()?;
       let header_len = header.encoded_len()?;
       let element_length = header.length;
```

This will also eliminate the need to use `self.position = pos;` to reset the internal position, as you can calculate the length of the TLV production using the above.